pkg/reconciler/revision/background.go

```diff
@@ -0,0 +1,258 @@
+	err       error
+	remaining int
+}
+
+// workItem is a single task submitted to the queue, to resolve a single image
+// for a resolveResult.
+type workItem struct {
+	result  *resolveResult
+	timeout time.Duration
+
+	name  string
+	image string
+	index int
+}
+
+func newBackgroundResolver(logger *zap.SugaredLogger, resolver imageResolver, enqueue func(types.NamespacedName)) *backgroundResolver {
+	r := &backgroundResolver{
+		logger: logger,
+
+		resolver: resolver,
+		enqueue:  enqueue,
+
+		results: make(map[types.NamespacedName]*resolveResult),
+		queue:   workqueue.NewNamed("digests"),
+	}
+
+	return r
+}
+
+// Start starts the worker threads and runs maxInFlight workers until the stop
+// channel is closed. It returns a done channel which will be closed when all
+// workers have exited.
+func (r *backgroundResolver) Start(stop <-chan struct{}, maxInFlight int) (done chan struct{}) {
+	var wg sync.WaitGroup
+
+	// Run the worker threads.
+	wg.Add(maxInFlight)
+	for i := 0; i < maxInFlight; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				item, shutdown := r.queue.Get()
+				if shutdown {
+					return
+				}
+
+				rrItem, ok := item.(*workItem)
+				if !ok {
+					r.logger.Fatalf("Unexpected work item type: want: %T, got: %T\n", &workItem{}, item)
+				}
+
+				r.processWorkItem(rrItem)
+			}
+		}()
+	}
+
+	// Shut down the queue once the stop channel is closed, this will cause all
+	// the workers to exit.
+	go func() {
+		<-stop
+		r.queue.ShutDown()
+	}()
+
+	// Return a done channel which is closed once all workers exit.
+	done = make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	return done
+}
+
+// Resolve is intended to be called from a reconciler to resolve a revision. If
+// the resolver already has the digest in cache it is returned immediately, if
+// it does not and no resolution is already in flight a resolution is triggered
+// in the background.
+// If this method returns `nil, nil` this implies a resolve was triggered or is
+// already in progress, so the reconciler should exit and wait for the revision
+// to be re-enqueued when the result is ready.
+func (r *backgroundResolver) Resolve(rev *v1.Revision, opt k8schain.Options, registriesToSkip sets.String, timeout time.Duration) ([]v1.ContainerStatus, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	name := types.NamespacedName{
+		Name:      rev.Name,
+		Namespace: rev.Namespace,
+	}
+
+	result, inFlight := r.results[name]
+	if !inFlight {
+		r.addWorkItems(rev, name, opt, registriesToSkip, timeout)
+		return nil, nil
+	}
+
+	if !result.ready() {
+		return nil, nil
+	}
+
+	ret := r.results[name]
+	return ret.statuses, ret.err
+}
+
+// addWorkItems adds a digest resolve item to the queue for each container in the revision.
+// This is expected to be called with the mutex locked.
+func (r *backgroundResolver) addWorkItems(rev *v1.Revision, name types.NamespacedName, opt k8schain.Options, registriesToSkip sets.String, timeout time.Duration) {
```

Comment:
Feels somewhat redundant to pass both revision and name down here.

Reply:
well, we need revision and it was [a previous review comment](https://github.com/knative/serving/pull/9354#discussion_r486408585) from @vagababov (that you thumbs-upped :D) that asked to pass name down since we have it in the caller :(

Reply:
oh well then :joy: